server/test/integration/apiTest/messageClientsTest.js

```diff
@@ -0,0 +1,198 @@
+      fallbackPhoneNumbers: [],
+      heartbeatPhoneNumbers: [],
+      fromPhoneNumber: testFromPhoneNumber,
+      isSendingAlerts: true,
+      isSendingVitals: true,
+    })
+    nonActiveClients.push(
+      await factories.clientDBFactory(db, {
+        displayName: "Test Client Don't Send Alerts",
+        responderPhoneNumbers: [],
+        fallbackPhoneNumbers: [testNotExpectPhoneNumbers[1]],
+        heartbeatPhoneNumbers: [],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: false,
+        isSendingVitals: true,
+      }),
+    )
+    nonActiveClients.push(
+      await factories.clientDBFactory(db, {
+        displayName: "Test Client Don't Send Vitals",
+        responderPhoneNumbers: [],
+        fallbackPhoneNumbers: [],
+        heartbeatPhoneNumbers: [testNotExpectPhoneNumbers[2]],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: true,
+        isSendingVitals: false,
+      }),
+    )
+    nonActiveClients.push(
+      await factories.clientDBFactory(db, {
+        displayName: "Test Client Don't Send",
+        responderPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        fallbackPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        heartbeatPhoneNumbers: [testNotExpectPhoneNumbers[3]],
+        fromPhoneNumber: testFromPhoneNumber,
+        isSendingAlerts: false,
+        isSendingVitals: false,
+      }),
+    )
+
+    // create locations for all non-active clients
+    for (const client of nonActiveClients) {
+      await locationDBFactory(db, {
+        locationid: `Location${client.id.substr(0, 8)}`,
+        displayName: `Location for ${client.displayName}`,
+        clientId: client.id,
+        isSendingAlerts: Math.round(Math.random()),
+        isSendingVitals: Math.round(Math.random()),
+      })
+    }
+
+    // stub twilioHelpers.sendTwilioMessage
+    sandbox.stub(twilioHelpers, 'sendTwilioMessage')
+    this.agent = chai.request.agent(server)
+  })
+
+  afterEach(async () => {
+    sandbox.restore()
+    await db.clearTables()
+    this.agent.close()
+  })
+
+  describe('for a good request', () => {
+    it('should respond with status 200 (OK)', async () => {
+      const response = await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      expect(response).to.have.status(200)
+    })
+
+    it('should message all active clients', async () => {
+      await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      testExpectPhoneNumbers.forEach(pn => {
+        expect(twilioHelpers.sendTwilioMessage).to.be.calledWith(pn, testFromPhoneNumber, message)
+      })
+    })
+
+    it('should not message any non-active clients', async () => {
+      await this.agent.post('/api/message-clients').send({ braveKey, message })
+
+      testNotExpectPhoneNumbers.forEach(pn => {
+        expect(twilioHelpers.sendTwilioMessage).to.not.be.calledWith(pn, testFromPhoneNumber, message)
+      })
+    })
+  })
```

Comment:
This is a good overall test - but if it fails it may be hard to determine which non-active client caused the failure, which doesn't really pinpoint the problem. That being said it would be more work to make more specific tests, so I'm also not sure it's worth the time/effort.

Reply:
This is definitely a wall that I'm hitting.

I've found that there are many errors to check for this API call; only some are covered:
- [x] a client that should be contacted isn't contacted
- [x] a client that shouldn't be contacted is contacted

However, if these fail, it doesn't provide any additional information about who caused the error, or how; it just outputs that it failed one of those checks.

There is also the case where Twilio fails to send a message, and you and Theresa mentioned that this should be logged. I'm going to implement some logic to keep track of failed messages and would appreciate your thoughts on that when I make the commit!

Reply:
Ya this one's tricky, I wonder if it would be worth testing that when a Twilio message fails to send that an error is in fact logged? I also wonder if it's worth testing what happens if there's no API Key/blank API key? These are just thoughts to consider though.

Reply:
Ah! Good idea with testing for no API key given! I'll add that as a test case.

Reply:
Awesome, I can't wait!

Reply:
Sounds good!